Mathlib/Data/Complex/Basic.lean

```diff
@@ -941,6 +941,16 @@ theorem im_eq_sub_conj (z : ℂ) : (z.im : ℂ) = (z - conj z) / (2 * I) := by
     mul_div_cancel_left₀ _ (mul_ne_zero two_ne_zero I_ne_zero : 2 * I ≠ 0)]
 #align complex.im_eq_sub_conj Complex.im_eq_sub_conj
 
+
+/-- Show the imaginary number ⟨x, y⟩ as an "x + y*I" string
+
+Note that the Real numbers used for x and y will show as cauchy sequences due to the way Real
+numbers are represented.
+-/
+unsafe instance repr : Repr Complex where
+  reprPrec f p := Repr.addAppParen (_root_.repr f.re ++ " + " ++ _root_.repr f.im ++ "*I") p
```

Comment:
Probably this should be conditional on the precedence of `+` (65) rather than "app" (1024)

Reply:
I am struggling to make this happen, partly because I cannot think of a way to test this. I have updated everything else based on your feedback, thanks for that.

Reply:
Maybe the easy way to test this would be to add a `Repr` for `DualNumber` at the same time, and check that `DualNumber Complex` has parentheses in the right place?

Reply:
I will work on that, good suggestion.

Reply:
I wrote this piece of code in `Algebra/DualNumber.lean`:

```
/-- Show DualNumber with values x and y as an "x + y*ε" string -/
instance instRepr [Repr R] : Repr (DualNumber R) where
  reprPrec f _p := repr f.fst ++ " + " ++ repr f.snd ++ "*ε"

```

I made Repr Complex print `p`, which was `0` for both cases (for a, b (both complex) in `a + bε`). This suggests to me that I am not getting sufficient information to conditionally print the brackets. I will for now opt to _always_ print parentheses. I checked for Repr Rat as well, and that one does not add any outside brackets even though it is represented as `a / b`, which I'd presume should get parentheses if squared, for example. 

Maybe I am missing something, but I tried finding other Repr's that handle conditionally adding brackets, and couldn't find anything. 

I will also add parentheses for DualNumbers, to keep them consistent.

Let me know if there is a way, of course, but I do not see it.

Reply:
This is a great question; the best I could come up with was
```lean
elab:max "prec_of% " t:term : term => do
  let ⟨.node _ name _⟩ := t | throwErrorAt t "Cannot find syntax node"
  let p ← unsafe Lean.evalConst Lean.ParserDescr name
  let .trailingNode _ prec _ _ := p | throwErrorAt t "Cannot find precedence"
  return Lean.toExpr prec

#eval prec_of% _ * _
```